zip-0316.rst

```diff
@@ -604,15 +613,84 @@ are reserved to indicate Metadata Items other than Receivers or
+
+As of `Revision 1`_, the subset of Metadata typecodes in the range
+:math:`mathtt{0xE0}` to :math:`mathtt{0xEF}` inclusive are designated
+as "MUST-understand": if a Consumer is unable to recognise the meaning
+of a Metadata Item with a typecode in this range, then it MUST regard
+the entire UA/UVK as unsupported and not process it further.
+
 Since Metadata Items are not Receivers, they MUST NOT be selected by
 a Sender when choosing a Receiver to send to, and since they are not
 Viewing Keys, they MUST NOT provide additional authority to view
 information about transactions.
 
-Currently no Metadata Types are defined. New Metadata Types SHOULD be
-introduced either by a modification to this ZIP or by a new ZIP, in
-accordance with the ZIP Process [#zip-0000]_.
+Address Expiration Metadata
+---------------------------
 
+As of `Revision 1`_, typecodes :math:`\mathtt{0xE0}` and :math:`\mathtt{0xE1}`
+are reserved for optional address expiry metadata. A producer MAY choose to
+generate Unified Addresses containing either or both of the following Metadata
+Item Types, or none.
+
+The value of a :math:`\mathtt{0xE0}` item MUST be an unsigned 32-bit integer in
+little-endian order specifying the Address Expiry Height, a block height of the
+Zcash chain associated with the UA/UVK. A Unified Address containing metadata
+typecode :math:`\mathtt{0xE0}` MUST be considered expired when the height of
+the Zcash chain is greater than this value.
+
+The value of a :math:`\mathtt{0xE1}` item MUST be an unsigned 64-bit integer in
+little-endian order specifying a Unix Epoch Time, hereafter referred to as the
+Address Expiry Time. A Unified Address containing metadata typecode
+:math:`\mathtt{0xE1}` MUST be considered expired when the current time is
+after the Address Expiry Time.
+
+A Sender that supports `Revision 1`_ of this specification MUST set
+a non-zero ``nExpiryHeight`` field in transactions that it creates. If the
+``nExpiryHeight`` normally constructed by the Sender would be greater than the
+Address Expiry Height, then the transaction MUST NOT be sent. If only an
+Address Expiry Time is specified, then the Sender MUST choose a value for
+``nExpiryHeight`` such that the transaction will expire no more than 24 hours
+after the current time.
+
+If wallet user attempts to send to an expired address, the error presented to
+the user by the wallet SHOULD include a suggestion that the user should
+attempt to obtain a currently-valid address for the intended recipient.
+
+When deriving from a UFVK or UIVK containing typecodes :math:`\mathtt{0xE0}`
+and/or :math:`\mathtt{0xE1}`, these Metadata Items MUST be retained unmodified
+in the output.
+
+Address expiration imposes no constraints on the Producer of an address. A
+Producer MAY generate multiple Unified Addresses with the same Receivers but
+different expiration metadata and/or any number of distinct Diversified Unified
+Addresses with the same or different expiry metadata, in any combination.
+
```

Comment:
```suggestion

Producers of Diversified Unified Addresses should be aware that the expiration
metadata could potentially be used to link addresses from the same source.
Normally, if Diversified Unified Addresses derived from the same UIVK contain
only Sapling and/or Orchard Receivers and no Metadata Items, they will be
unlinkable as described in [#protocol-concretediversifyhash]_; this property
does not hold when Metadata Items are present.

```
Also add this reference:
```rst
.. [#protocol-concretediversifyhash] `Zcash Protocol Specification, Version 2023.4.0. Section 5.4.1.6: DiversifyHash^Sapling and DiversifyHash^Orchard Hash Functions <protocol/protocol.pdf#concretediversifyhash>`_
```
(Optionally update the version for the other protocol spec references to 2023.4.0 as well. I have checked that none of the referenced sections were renumbered.)

Reply:
What we should do instead is, for address derivation, specify that the UIVK expiry height and time MUST be treated as upper bounds, and recommend that expiry heights and times be selected in a fashion that avoids this sort of linkability.